[image-picker] Rewrite ncl screen

bbarthec opened the pull request:
# Why

Resolves [ENG-3077](https://linear.app/expo/issue/ENG-3077/audit-expo-image-picker-examples-screen)

# How

I've modified the `Configurator` idea proposed in #16267:
- `FunctionDemo` component is created to fully illustrate single function API with possibility to adjust every argument:
  - boolean -> checkbox
  - string/number/enum -> cyclic button over list of possible values
  - object -> each property is handled individually

# Demo

https://user-images.githubusercontent.com/16623003/155548871-fc6b16c3-8f71-4fb7-a6cf-4d94253c307b.mp4

# Test Plan

- Open `ncl` and navigate to `WebBrowser` or `ImagePicker` screen to play with the new component.

# TODOs

- [x] Trim long result parts (like `base64` property - [`text-overflow: ellipsis`](https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow)
- [x] Add/restore support for `platforms`
- [x] Restore commented out entries - mapping one type (boolean) to other type (color string)
- [x] test Android & Web

bbarthec committed:
[blur-view] Move `useResettingState` to the upper level utils & refactor `useLoopingAnimatedValue`

bbarthec committed:
[image-picker] Moce `ImagePickerScreen` into the separate directory

bbarthec committed:
[ncl] Create `DisappearingMonoText`

bbarthec committed:
[ncl][image-picker] Refactor demos for permission methods

bbarthec committed:
[ncl] Create `FunctionDemo` component

Create generic component for displaying and configuraing functions exposed by the module

bbarthec committed:
[imge-picker][ncl] Use `FunctionDemo` component

Migrate `image-picker`'s screen to `FunctionDemo` component

bbarthec committed:
Rename `DisappearingMonoText` to `MonoTextWithCountdown`

bbarthec committed:
Add `constant` parameter type, some comments and `additionalParameters` prop

bbarthec committed:
Adjust ImagePicker screen

bbarthec committed:
Rewrite `OpenAuthSessionAsyncDemo` to use the `FunctionDemo` component

bbarthec committed:
Rewrite `OpenBrowserAsyncDemo` to use `FunctionDemo` component

- some properties are commented out as I need to came up with an idea how to handle them

bbarthec committed:
[ncl] Add platforms support for `FunctionDemo` component

bbarthec committed:
Add trimming to very long string results

bbarthec committed:
[ncl] Add support for colors using enum-like structure

Restore commented out colors parameters

bbarthec committed:
Restore initialState for unsupported parameters on given platform

bbarthec commented:
There's a lot of redundancy on the screen right now 🤔 
I'm not sure how we should design section for each method, but I would like to have it this way.
We can add some more mechanisms as well (parameters presets, logging some info in the console, measuring execution time, etc).

barthap reviewed (approved):
Looks great! 👏

tsapeta commented on apps/native-component-list/src/screens/WebBrowser/OpenAuthSessionAsyncDemo.tsx:
```diff
@@ -1,60 +1,41 @@
+    <FunctionDemo
+      namespace="WebBrowser"
+      name="openAuthSessionAsync"
+      parameters={[
+        {
+          name: 'url',
+          type: 'constant',
+          value: 'url',
+        },
+        { name: 'redirectUrl', type: 'constant', value: redirectUrl },
+        {
+          name: 'options',
+          type: 'object',
+          properties: [{ name: 'createTask', type: 'boolean', initial: true }],
+        },
+      ]}
+      additionalParameters={[{ name: 'shouldPrompt', type: 'boolean', initial: false }]}
+      actions={(
+        _: string,
+        redirectUrl: string,
+        options: WebBrowser.WebBrowserOpenOptions,
+        shouldPrompt: boolean
+      ) => {
+        const url = `https://fake-auth.netlify.com?state=faker&redirect_uri=${encodeURIComponent(
+          redirectUrl
+        )}&prompt=${shouldPrompt ? 'consent' : 'none'}`;
+        return WebBrowser.openAuthSessionAsync(url, redirectUrl, options);
+      }}
```
Why do you put so many things into the render function? It becomes harder to read, less concise and less efficient.

bbarthec replied:
Yeah! I've refactored and cleaned up the whole PR now, so give it another shot! 😁

tsapeta commented on apps/native-component-list/src/screens/WebBrowser/OpenBrowserAsyncDemo.tsx:
```diff
@@ -1,111 +1,116 @@
+    <FunctionDemo
+      namespace="WebBrowser"
+      name="openBrowserAsync"
+      parameters={[
```
Again, it looks terrible to have so big and static object in the render function

tsapeta commented on apps/native-component-list/src/components/MonoTextWithCountdown.tsx:
```diff
@@ -0,0 +1,127 @@
+  onCountdownEnded: () => void;
+}>;
+
+function MonoTextWithCountdown({ children, timeout = 8000, onCountdownEnded }: Props) {
```
Could you explain the reason behind having the disappearing, pausable and resumable component? I don't see any benefit of this 😅

bbarthec replied:
When I was working with the old `ImagePicker` demo screen I was wondering why every result from the function call is presented differently (some additional section on the screen, alert dialog, etc.).
As almost every function in our APIs is returning some JSON-like object I started to wonder about the single unified way to present these results.
`MonoText` addresses this problem perfectly, because the results are clearly rendered on the screen.
But there's a problem with the `MonoText` though 😞 
It takes so much space on the screen even when the result could be rendered as a single line.
Most of the time results could be verified in just a few moments so I decided I should autohide `MonoText` by default, because it clutters the screen on the long run (especially if the screen presents more than one `FunctionDemo` - e.g. `ImagePicker` case).
But there're cases when you want to investigate the returned result more thoroughly and then autohiding after a fixed timeframe seems wrong, so I decided to add pause mechanism (resume mechanism was trivial at this point so I've added it as well) 😁
And that's the full story 😉

tsapeta commented on apps/native-component-list/src/components/FunctionConfigurator.tsx:
```diff
@@ -0,0 +1,212 @@
+  onChange: (name: string | [string, string], newValue: PrimitiveArgument) => void;
+};
+
+export default function FunctionConfigurator({
```
The idea behind `Configurator` was to do simple thing: return an object with options based on given config, without knowing what to do with the object and how it's going to be used, so I think the `Function` prefix is not necessary here.

bbarthec replied:
I've modified the `Configurator` and at this point it operates on these fixed types (that are very sophisticated at the first glimpse) that are directly connected with the `FunctionDemo`. I've changed the name to the `Configurator`, but I wonder what you think now 🤔

bbarthec committed:
Refactor `FunctionDemo` component

- split out every component into a separate file
- clean-up the code
- add comments for the parts that are not clear at the first glance
- refactor `FunctionSignature` component
- renamed `FunctionConfigurator` into `Configurator`

bbarthec committed:
[image-picker] Extracted `ImageOrVideoPreview` component into separate file

bbarthec committed:
[web-browser] Extracted static demo components configurations from the render functions

bbarthec committed:
Add `ArgumentName` and `OnArgumentChangeCallback` types

bbarthec committed:
Remove `console.log`

bbarthec commented on apps/native-component-list/src/components/FunctionDemo.types.ts:
```diff
@@ -0,0 +1,51 @@
+export type Tuple = [number, number];
```
That's the case I've encountered so far, but I've added a comment stating that this type is extendable if needed 👀

bbarthec commented:
I've just updated the demo recording 👀

barthap reviewed (approved):
I'm a simple man. It works, it looks nice, I like it.

Simek commented on apps/native-component-list/src/components/FunctionDemo/EnumButton.tsx:
```diff
@@ -0,0 +1,69 @@
+  const valuesAreEnums = useEnumValues(values);
+
+  const handleOnPress = useCallback(() => {
+    const newValue = valuesAreEnums
+      ? values[(values.findIndex((v) => v.value === value) + 1) % values.length].value
+      : values[(values.findIndex((v) => v === value) + 1) % values.length];
```
```suggestion
    const newValue = values.map(v => valuesAreEnums ? v.value : v).find((v) => v === value);
```

Just a small simplification suggestion, without the checkout, hope I understood correctly what the code should do. 🙂

bbarthec committed:
Simplify methods

bbarthec committed:
[no ci] Remove redundant blank lines
